src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs

```diff
@@ -893,7 +912,47 @@ private ZipArchiveEntry CreateEntry(ZipArchive package, string entryName, Compre
         private ZipArchiveEntry CreatePackageFileEntry(ZipArchive package, string entryName, DateTimeOffset timeOffset, CompressionLevel compressionLevel)
         {
             var entry = package.CreateEntry(entryName, compressionLevel);
-            entry.LastWriteTime = timeOffset;
+
+            // Please note: ZipArchive stream reader sometime changes LastWriteTime by another 1 second off than what "entry.LastWriteTime" has.
+            // The FAT filesystem of DOS has a timestamp resolution of only two seconds; ZIP file records mimic this.
+            // As a result, the built -in timestamp resolution of files in a ZIP archive is only two seconds, though extra fields can be used to store more precise timestamps.The ZIP format has no notion of time zone, so timestamps are only meaningful if it is known what time zone they were created in.
+            if (timeOffset.UtcDateTime < ZipFormatMinDate)
+            {
+                if (!_zipFormatCorrected)
+                {
+                    _zipFormatCorrected = true;
+
+                    _logger.Log(
+                        PackagingLogMessage.CreateMessage(
+                            string.Format(
+                                CultureInfo.CurrentCulture,
+                                Strings.ZipFileTimeStampModified, entryName, timeOffset.DateTime.ToShortDateString(), ZipFormatMinDate.ToShortDateString()),
+                            LogLevel.Information));
+                }
+
+                entry.LastWriteTime = ZipFormatMinDate;
+            }
+            else if (timeOffset.UtcDateTime > ZipFormatMaxDate)
+            {
+                if (!_zipFormatCorrected)
+                {
+                    _zipFormatCorrected = true;
+
+                    _logger.Log(
+                        PackagingLogMessage.CreateMessage(
+                            string.Format(
+                                CultureInfo.CurrentCulture,
+                                Strings.ZipFileTimeStampModified, entryName, timeOffset.DateTime.ToShortDateString(), ZipFormatMinDate.ToShortDateString()),
+                            LogLevel.Information));
+                }
+
+                entry.LastWriteTime = ZipFormatMaxDate;
+            }
+            else
+            {
+                entry.LastWriteTime = timeOffset;
```

Comment:
Given the problems we had when deterministic pack was briefly enabled, I don't think this is writing the correct time. I think we need to explicitly convert it to UTC, and in the package extraction code make sure we tell the System.IO API that we're writing UTC, or convert to locally.

Reply:
I want to encourage @erdembayar to read through the comments in the issue, in particular https://github.com/NuGet/Home/issues/7001#issuecomment-534215527 and below. 

Basically people would try to repack assemblies that came from packages that were built with the deterministic switch, with the time set at 1980, and would fail.

Reply:
It's already writing UTC time, because this one is coming from `file.GetStream()`

https://github.com/NuGet/NuGet.Client/blob/c920e219edca36f265c267f4e314adfcf8d47bd4/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs#L922

In turn it get UTC datetime of file:

https://github.com/NuGet/NuGet.Client/blob/c920e219edca36f265c267f4e314adfcf8d47bd4/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PhysicalPackageFile.cs#L96-L101

Next it calls my changes with UTC datetime:

https://github.com/NuGet/NuGet.Client/blob/c920e219edca36f265c267f4e314adfcf8d47bd4/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs#L926-L930

Reply:
@zivkan 
Addressed. I tested on PST and Samao timezones.